Comment on tests?

carolyngreeno opened the pull request:
I’m not sure I’m writing the tests correctly. I wrote the first couple 
with my tutor.

I think I’m still struggling with the concept of testing. I could use
some help figuring out what to test for and then once I do, with any
luck, I can figure out the right code to do it.

I’m not trying to do any validation on the user input right? This is
strictly testing my own code for weak points. For example, for
getDayName, would I want to create a var for a string like “Thursday”
and then compare it to a date in a format like “May 14 2015” if that
day is a Thursday?

Commit:
Comment on tests?

I’m not sure I’m writing the tests correctly. I wrote the first couple
with my tutor.

I think I’m still struggling with the concept of testing. I could use
some help figuring out what to test for and then once I do, with any
luck, I can figure out the right code to do it.

I’m not trying to do any validation on the user input right? This is
strictly testing my own code for weak points. For example, for
getDayName, would I want to create a var for a string like “Thursday”
and then compare it to a date in a format like “May 14 2015” if that
day is a Thursday?

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+	// setDate
+	describe("setDate", function () {
+
+		it("Return same date object as date object given.", function () {
```
The test title suggests something a little different than what the test itself tests. If you want to test the object returned, you can test that the object given is deep.equal to the object returned. You can also test the result of calling toString on the objects, but it's an additional step beyond testing the method itself. For example:
`expect(enhancedDate.setDate(newDate)).to.deep.equal(newDate);`

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+			var setDateResult = enhancedDate.setDate(dateNumber);
+			var setDateNumber = setDateResult.valueOf();
+
+			expect(dateNumber).to.equal(setDateNumber);
```
This test works as well, but I would consider testing the method more directly. For example:
`expect(enhancedDate.setDate(dateNumber)).to.deep.equal(new Date(dateNumber);`

carolyngreeno replied:
I’ve read the comments and I’m working my little brain as hard as I can. I used this example but I can’t quite connect the dots as to _how_ it’s work. Can you tell me how the two objects(?) are being compared? What’s the difference between them?

To me it look like the first one is calling setDate method from the enhancedDate module and using the dateNumber of milliseconds supplied and comparing it to the plain old date object (not referencing the enhancedDate module) using the same number of milliseconds. If that’s right then were using a value we enter to compare it to a value returned by the enhancedDate module? Am I in the ballpark at all? Did that even make sense?

it(“Description here.", function () {
    var dateNumber = 1432406605;
    expect(enhancedDate.setDate(dateNumber)).to.deep.equal(new Date(dateNumber));
});

> On May 24, 2015, at 12:40 PM, jamdon notifications@github.com wrote:
> 
> In js/tests/enhancedDate.spec.js:
> 
> > - > +   it("Return same date object as date object given.", function () { > +   var newDate = new Date("Dec 12, 2015"); > + var newDateString = newDate.toString(); > + var setDateResult = enhancedDate.setDate(newDate); > +  var setDateString = setDateResult.toString(); > + > + expect(newDateString).to.equal(setDateString); > +    }); > + > + it("Give unix time then get date object back that represents that same time.", function () { > +    var dateNumber = 1432406605; > +    var setDateResult = enhancedDate.setDate(dateNumber); > +   var setDateNumber = setDateResult.valueOf(); > + > +    expect(dateNumber).to.equal(setDateNumber); 
> >   This test works as well, but I would consider testing the method more directly. For example:
> >   expect(enhancedDate.setDate(dateNumber)).to.deep.equal(new Date(dateNumber);
> 
> —
> Reply to this email directly or view it on GitHub.

jamdon replied:
Yes, that's correct. The setDate method is saying it will do something, and we're validating that it will do what it says it will. In this case, if we provide setDate a millisecond value, it should be creating a new Date object from that value and returning it. We validate this is what happens by creating a new native Date object using that value and validating the two create identical objects.

carolyngreeno replied:
Can you tell me if I’m on the right track with testing booleans for isToday?

// isToday
describe("isToday", function () {

   it("Test should return true.", function () {
      var setToday = true;

```
  expect(enhancedDate.isToday()).to.equal(setToday);
```

   });

   it("Test should return false.", function () {
      var setToday = false;

```
  expect(enhancedDate.isToday()).to.equal(setToday);
```

   });
});

> On May 24, 2015, at 3:14 PM, jamdon notifications@github.com wrote:
> 
> In js/tests/enhancedDate.spec.js:
> 
> > +
> > -       it("Return same date object as date object given.", function () {
> > -           var newDate = new Date("Dec 12, 2015");
> > -           var newDateString = newDate.toString();
> > -           var setDateResult = enhancedDate.setDate(newDate);
> > -           var setDateString = setDateResult.toString();
> >   +
> > -           expect(newDateString).to.equal(setDateString);
> > -       });
> >   +
> > -       it("Give unix time then get date object back that represents that same time.", function () {
> > -           var dateNumber = 1432406605;
> > -           var setDateResult = enhancedDate.setDate(dateNumber);
> > -           var setDateNumber = setDateResult.valueOf();
> >   +
> > -           expect(dateNumber).to.equal(setDateNumber);
> 
> Yes, that's correct. The setDate method is saying it will do something, and we're validating that it will do what it says it will. In this case, if we provide setDate a millisecond value, it should be creating a new Date object from that value and returning it. We validate this is what happens by creating a new native Date object using that value and validating the two create identical objects.
> 
> —
> Reply to this email directly or view it on GitHub.

jamdon replied:
Chai has an easier way of validating true or false: 

`expect(enhancedDate.isToday()).to.be.true;`

or

`expect(enhancedDate.isToday()).to.be.false;`

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+	// Test each method within object
+
+	// setDate
+	describe("setDate", function () {
```
One more obvious test is to simply test that the return value is of the expected type:
`expect(enhancedDate.setDate()).is.an.instanceof('Date');`

jamdon replied:
One basic function I'd consider is simply checking the type returned by setDate:
`expect(enhancedDate.setDate()).to.be.an.instanceof(Date);`

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+	});
+
+	// getDate
+	describe("getDate", function () {
```
getDate has two types of return values, so I would write tests that validates each return type.

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+		it("Create a new Date object to set date with.", function () {
+			var dayName = enhancedDate(new Date(2015, 4, 14));
+
+			expect(dayName).to.equal("Thursday");
```
Additional tests you could write include testing other possible outputs or testing the return type.

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+	// getMonthName
+	describe("getMonthName", function () {
+
+		it("", function () {
```
Same deal with getMonthName as getDayName

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+	describe("isToday", function () {
+
+		it("", function () {
+
```
Because this is a boolean method, you can test one case where it should return true and once case where it should return false.

jamdon commented on js/tests/enhancedDate.spec.js:
```diff
@@ -0,0 +1,69 @@
+	describe("isFuture", function () {
+
+		it("", function () {
+
```
same deal here as with isToday - one test for the true case and one for the false

jamdon commented:
Unit tests aren't for testing input validation. In this case they are specifically for testing the methods exposed on the enhancedDate module. Each of the methods returns a value, so you'll want to test that the value returned is what is expected for a given input or initial state. For some of these functions, there are two possible return values (functions that return true or false - test both cases.) For two of the functions, getDayName and getMonthName, there are multiple return values that can be tested. For any of the methods, you can test the type of the return value. In general, unit tests are very straightforward an obvious, which clearly expected results that you can predict. Does that help?

jamdon commented:
Additional comment inline.

Commit:
final commit

Here are my final changes after watching your video. It’s good to know
I was making it way too complicated. I think I have a better feel for
what to do now. I have one question about testing isFuture. If the date
is today, it’s also not in the future. Do we not test for that in
isFuture because we’re already testing for it isToday?

jamdon commented:
Technically a given date can be in the future and still be "today" when the test is run. If, for example, you set a date one hour in the future, it could still be today. So isFuture just needs to do a simple comparison to see which value is greater, while isToday needs to check three different things being equal to make sure it's "today".

jamdon commented:
:+1:
